File: 2601/prime-subtraction-operation2.py
# LeetCode 2601. Prime Subtraction Operation
# 有一堆數，每次「挑1個」沒挑過的數，把它減掉更小的「質數」
# 減多次後，能不能讓 nums[i] 變「嚴格遞增數列」？
class Solution:
    def primeSubOperation(self, nums: List[int]) -> bool:
        prime = []  # 使用「篩子法」，收集 1000 以下的質數
        possible = [True] * 1001  # 可能是質數
        for i in range(2,1001):
            if possible[i]: 
                prime.append(i)  # 收集到質數
                for k in range(i*i, 1001, i): possible[k] = False  # 刪掉它的倍數
        def findPrimeBelow(now):
            index = bisect_left(prime, now) - 1 # 找到左邊的插入點，-1 就是「更小的值」
            if index==-1: return 0  # 沒有適合的質數
            else: return prime[index]  # 更小的質數
        
        nums[0] -= findPrimeBelow(nums[0])  # 最前面的數，先壓到最小
        for i in range(1, len(nums)):  # 剩下的數，要慢慢增加
            nums[i] -= findPrimeBelow(nums[i] - nums[i-1])  # 將「增加的幅度」壓到最小（但還是會漸增）
            if nums[i-1] >= nums[i]: return False  # 壓完後，無法成功，就失敗了
        return True  # 沒有失敗，就成功了！

File: 2601/prime-subtraction-operation2.cpp
// LeetCode 2601. Prime Subtraction Operation
class Solution {
public:
    vector<int> prime;
    void buildPrime() {
        int table[1001] = {};
        for(int i=2; i<=1000; i++) {
            if(table[i]==0){
                prime.push_back(i);
                for(int k=i*i; k<=1000; k+=i) table[k] = 1;
            }
        }
    }
    int findPrimeBelow(int now) {
        auto pos = upper_bound(prime.begin(), prime.end(), now-1);
        if(pos==prime.begin()) return 0;
        return prime[pos-prime.begin()-1];
    }
    bool primeSubOperation(vector<int>& nums) {
        if(prime.size()==0) buildPrime(); // 第1次使用時，建立 prime 陣列
        nums[0] -= findPrimeBelow(nums[0]);
        for(int i=1; i<nums.size(); i++) {
            nums[i] -= findPrimeBelow(nums[i]-nums[i-1]);
            if(nums[i]<=nums[i-1]) return false;
        }
        return true;
    }
};
